torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -0,0 +1,384 @@
+    default:                                           \
+      throw std::runtime_error("Invalid scalar type"); \
+  }
+
+namespace c10d {
+
+using KeyType = AlgorithmKey;
+using EntryType = std::unique_ptr<AlgorithmEntry>;
+
+namespace {
+
+// Wrap c10d store as Gloo store
+class GlooStore : public ::gloo::rendezvous::Store {
+ public:
+  GlooStore(const std::shared_ptr<::c10d::Store>& store) : store_(store) {}
+
+  void set(const std::string& key, const std::vector<char>& value) override {
+    std::vector<uint8_t> tmp(value.begin(), value.end());
+    store_->set(key, tmp);
+  }
+
+  std::vector<char> get(const std::string& key) override {
+    auto value = store_->get(key);
+    return std::vector<char>(value.begin(), value.end());
+  }
+
+  void wait(const std::vector<std::string>& keys) override {
+    store_->wait(keys, Store::kDefaultTimeout);
+  }
+
+  void wait(
+      const std::vector<std::string>& keys,
+      const std::chrono::milliseconds& timeout) override {
+    store_->wait(keys, timeout);
+  }
+
+ protected:
+  std::shared_ptr<::c10d::Store> store_;
+};
+
+template <typename T>
+const ::gloo::ReductionFunction<T>* reductionFunction(const ReduceOp& r) {
+  switch (r) {
+    case ReduceOp::SUM:
+      return ::gloo::ReductionFunction<T>::sum;
+    case ReduceOp::PRODUCT:
+      return ::gloo::ReductionFunction<T>::product;
+    case ReduceOp::MIN:
+      return ::gloo::ReductionFunction<T>::min;
+    case ReduceOp::MAX:
+      return ::gloo::ReductionFunction<T>::max;
+  }
+
+  throw std::runtime_error("Unhandled ReduceOp");
+}
+
+} // namespace
+
+ProcessGroupGloo::WorkGloo::WorkGloo() : completed_(false) {}
+
+ProcessGroupGloo::WorkGloo::~WorkGloo() {}
+
+bool ProcessGroupGloo::WorkGloo::isCompleted() const {
+  return completed_;
+}
+
+bool ProcessGroupGloo::WorkGloo::isSuccess() const {
+  return !ex_;
+}
+
+bool ProcessGroupGloo::WorkGloo::wait() {
+  std::unique_lock<std::mutex> lock(m_);
+  while (!completed_) {
+    cv_.wait(lock);
+  }
+  return isSuccess();
+}
+
+const std::exception& ProcessGroupGloo::WorkGloo::exception() const {
+  return *ex_;
+}
+
+void ProcessGroupGloo::WorkGloo::finish() {
+  {
+    std::unique_lock<std::mutex> lock(m_);
+    completed_ = true;
+  }
+  cv_.notify_all();
+}
+
+void ProcessGroupGloo::WorkGloo::finishWithException(
+    const ::gloo::Exception& ex) {
+  {
+    std::unique_lock<std::mutex> lock(m_);
+    completed_ = true;
+    ex_ = std::unique_ptr<::gloo::Exception>(new ::gloo::Exception(ex));
+  }
+  cv_.notify_all();
+}
+
+ProcessGroupGloo::Options::Options()
+    : timeout(std::chrono::milliseconds(10 * 1000)), threads(2) {}
+
+ProcessGroupGloo::ProcessGroupGloo(
+    const std::shared_ptr<Store>& store,
+    int rank,
+    int size,
+    Options options)
+    : ProcessGroup(rank, size), store_(new GlooStore(store)), stop_(false) {
+  auto& devices = options.devices;
+  if (devices.empty()) {
+    devices.push_back(::gloo::transport::tcp::CreateDevice("localhost"));
```

Comment:
Can we also add Infiniband detection here and automatically use that device if available? Like what we did in THD?

Reply:
For sure! Let's do this when we get to the IB tests etc.